### ISSUE TITLE
Query the number of formula terms

### ISSUE THREAD

**Issue by DavidWalz** [matthewwardrop/formulaic#107]
Hi @matthewwardrop, 
prior to 0.4 the terms of a Formula were a list and its length could be queried. Now Formula is an iterable. What would you say about implementing a `__len__` method for Formula?

E.g.
```
def __len__(self):
    return sum(1 for _ in self) 
```

**Comment by matthewwardrop** [comment]
Hi @DavidWalz ,

Thanks for your thorough nits, as always.

I'm not actually sure what I think about this. The length of the "formula" often has no bearing on the number of terms, for example: `len(Formula('y ~ x')) == 2`, because this is a "structured" formula with ordered top-level nodes of "lhs" and "rhs". Do you have an example where the length of a formula would be helpful?

**Comment by DavidWalz** [comment]
The number of lhs terms equals the number of model parameters. In our design of experiments tool we use this to determine the minimium required number of experiments before going into the optimization, see https://github.com/basf/doe/blob/29666ecc3a42af9cf3f7346d41c38abfeebaed1d/doe/design.py#L101

We can of course count the terms ourselves, but I was wondering if it wouldn't make sense to have this built in as a way to determine the the number of columns of the model matrix before it's constructed.

**Comment by matthewwardrop** [comment]
That makes a lot of sense. In eeneral, though, the model matrix output columns also depends on the data; for example, if you have categorical fields, the number of columns generated will depend on the columns present during initial model matrix generation (subsequent uses of the model-spec will have fixed shape). For your use-case, this is obviously not a problem.

I took a closer look at the code, and I did implement __len__ on `Structured` that always looks at the length of the `._structure` dictionary, which is inconsistent with how iteration works... I'll patch that to be consistent, which should also solve your problem. It might be worth checking the `._has_structure` attribute before using a simple `len()` in your case :).

P.S. Doe looks really interesting. I don't fully understand what it is doing yet, but given that I work in the experimentation space, I want to :). I'll look into it more!
